### ISSUE TITLE
Strawman for component-level testbed

### ISSUE THREAD

**Issue by yjbanov** [angular/angular#1146]
It is very common for application developers to test their components with their templates. Internally we call these tests "component tests" (as opposed to unit- or integration-tests). These tests are not unit tests as they exercise more than just the component class. Nor are they integration tests as far as the application is concerned; that term is reserved for testing cross-component interactions, full application tests, and tests that involve the back-end.

Here are the requirements:

* instantiate a component in a test along with its template and child components
* replace child components with mocks
* pass bound values to the component
* trigger change detection
* fake the event loop, a la [FakeAsync](https://github.com/google/quiver-dart/blob/cf72b438560a502ebb319e764700e96c1f4dbe01/lib/testing/src/async/fake_async.dart)
* verify that component implement change detection correctly (e.g. via observables, immutables)
* access underlying DOM to trigger user events, such as click on buttons (this is usually done with the assistance of the Page Objects framework)

Strawman:

```dart

// This annotation tells Angular about test-specific dependency overrides
@ComponentTest(
  component: MyComponent,
  dependencyOverrides: const {
    ChildComponent: MockChildComponent,
    RealGoogleMap: MockGoogleMap,
  }
)
main() {
  describe("my-component", () {
    TestBed<MyComponent> testBed;

    beforeEach(() {
      testBed = createTestBedFor(MyComponent);
    });

    it("should work as intended", fakeAsync(async) {
      // instantiates a new component
      MyComponent cmp = testBed.component();

      // pass bindings using normal setters
      cmp.field1 = "hello";
      cmp.field2 = "world";

      // detectChanges runs change detection and returns how many
      // components ended up reevaluated, allowing testing correctness
      // of change detection
      expect(testBed.detectChanges().length, 5);

      // emulate passage of time
      async.flushMicrotasks();
      async.elapse(seconds: 1);

      // get access to the DOM
      Element rootElem = testBed.element();
      rootElem.querySelector('some-button').click();

      expect(cmp.buttonClicked, true);
    });
  });
}
```

**Comment by matanlurey** [comment]
Apparently this may be partially covered by #615. But I'd like to keep this open if possible because the strawman above is what would ease migration.

**Comment by yjbanov** [comment]
related: https://docs.google.com/document/d/1qFfHwYCOLpLuTap_l6TCEaU4I0C-pVFVbTl2o4-XbUo/edit#heading=h.aetfsm7ib2fe

**Comment by naomiblack** [comment]
@vicb can you resolve this along with #615?

**Comment by mhevery** [comment]
@yjbanov I think we need to discuss this. I think this should be done through `TestBed` API rather than annotation. I am going to move it to later milestone for now.

**Comment by yjbanov** [comment]
@mhevery What would be the benefits of using imperative API over annotations?

**Comment by mhevery** [comment]
Annotations are static. Imperative APIs can be easily extended, conditionally applied etc in the tests and even in the middle of the test.

**Comment by yjbanov** [comment]
OTOH, imperative API shuts the door to any potential code generated approaches to DI as no static analysis can be done on it. Could you give any concrete examples that annotations would not support?

**Comment by mhevery** [comment]
Not sure code generation is high priority for tests.

**Comment by yjbanov** [comment]
It is a high priority for those paranoid (like myself) users who like to run their tests in an environment as close to production as possible. The code of reflection-based DI, compilation and change detection is different from that of codegen-based ones. You'd have to trust that the two behave identically, which is rarely the case and a source the hard-to-debug bugs.

**Comment by jbdeboer** [comment]
@yjbanov @naomiblack Could we sketch out what can and should be done for M7 with this issue.

It is important to GT and we are overlapping work at the moment.

**Comment by mhevery** [comment]
I think most of this is already covered here: https://github.com/angular/angular/blob/master/modules/angular2/src/test_lib/test_bed.js Not sure there is a pressing issue to be solved for the GT milestone.

**Comment by yjbanov** [comment]
@mhevery the existing test bed API is based on internal classes and lacks integration with the event loop (specifically microtasks) and change detection.

@vicb what do you think of the above proposal? We need to start working on this asap. If you are blocked on something urgent, please  we can reassign.

I'm ok with dropping the static API for the short term solution.

**Comment by vicb** [comment]
@yjbanov I think most of the requested features are available (`fakeAsync` excepted).

The priority right now should be to make the TestBed public (#1609) - might be a good idea to re-assign this item to someone with more bandwidth & who knows the render architecture better than I do.

I have been working a lot on zones lately (#1609). Once we come with a working solution for microtasks in JS/Dart (hopefully sometime this week), I'll resume the work on `fakeAsync()` _(as of today JS does not know what a microtask is which make it impossible to `microleap() / flushMicrotasks()`)_

**Comment by matanlurey** [comment]
I am looking into how Yegor's proposal could work, focusing on Dart. Will report back with any blockers.

**Comment by vicb** [comment]
@matanlurey I'm working on `fakeAsync` ATM (making sure we don't duplicate the effort)

**Comment by matanlurey** [comment]
I am just using the Dart (quiver/testing/async) library, which can be replaced by yours when it's ready.

**Comment by vicb** [comment]
My plan is to use [fake_async (Dart)](https://pub.dartlang.org/packages/fake_async) and [lolex (JS)](https://github.com/sinonjs/lolex) and come with Dart & JS wrappers. 

I think it will be easier now that JS zone also has a `scheduleMicrotask`

**Comment by mhevery** [comment]
I will work on this.

**Comment by yjbanov** [comment]
This has been superseded by https://github.com/angular/angular/issues/1812. Closing.
